File: backend/src/domain/models/work_order.py
from dataclasses import dataclass, field
from datetime import date, datetime
from enum import Enum
from uuid import UUID

from .technical_location import TechnicalLocation


class WorkOrderStatus(str, Enum):
    released = "released"
    notified = "notified"


class ControlKey(str, Enum):
    PMNN = "PMNN"
    PM01 = "PM01"


class DeviationKey(str, Enum):
    executed = "PM01 Executed"
    not_executed = "PM01 Not Executed"


class OrderType(str, Enum):
    OE01 = "OE01"
    OE02 = "OE02"
    OE03 = "OE03"
    OE04 = "OE04"


class PlannerGroup(str, Enum):
    mechanical = "mechanical"
    electrical = "electrical"
    electronic = "electronic"


class InstallationState(str, Enum):
    running = "running"
    stopped = "stopped"


@dataclass
class StepClosure:
    id: UUID
    step_id: UUID
    actual_duration: float
    deviation_key: DeviationKey
    work_description: str
    safety_question_response: bool
    submitted_at: datetime


@dataclass
class OTStep:
    id: UUID
    work_order_id: UUID
    position: int
    description: str
    control_key: ControlKey
    is_fixed: bool
    planned_intervention_time: float | None = None
    closure: StepClosure | None = None


@dataclass
class WorkOrder:
    id: UUID
    order_type: OrderType
    technical_location: TechnicalLocation
    assigned_technician_id: UUID
    created_by_id: UUID
    planner_group: PlannerGroup
    activity_class: str
    installation_state: InstallationState
    planned_start: date
    planned_end: date
    priority: int
    status: WorkOrderStatus
    notif_final: bool
    sin_ttbjo_real: bool
    shift_number: int
    created_at: datetime
    steps: list[OTStep] = field(default_factory=list)


@dataclass
class TechnicianWorkload:
    technician_id: UUID
    technician_name: str
    ot_count: int
    shift_number: int


class NoPm01StepError(Exception):
    pass


class StepAlreadyClosedError(Exception):
    pass


class OtNotAssignedError(Exception):
    pass

File: mobile/lib/domain/models/work_order.dart
import 'ot_step.dart';
import 'technical_location.dart';

enum WorkOrderStatus { released, notified }

class WorkOrder {
  const WorkOrder({
    required this.id,
    required this.orderType,
    required this.technicalLocation,
    required this.assignedTechnicianId,
    required this.priority,
    required this.plannedStart,
    required this.plannedEnd,
    required this.status,
    required this.shiftNumber,
    this.steps = const [],
    this.createdById,
    this.plannerGroup,
    this.activityClass,
    this.installationState,
    this.notifFinal = false,
    this.sinTtbjoReal = false,
    this.createdAt,
  });

  final String id;
  final String orderType;
  final TechnicalLocation technicalLocation;
  final String assignedTechnicianId;
  final int priority;
  final DateTime plannedStart;
  final DateTime plannedEnd;
  final WorkOrderStatus status;
  final int shiftNumber;
  final List<OTStep> steps;
  final String? createdById;
  final String? plannerGroup;
  final String? activityClass;
  final String? installationState;
  final bool notifFinal;
  final bool sinTtbjoReal;
  final DateTime? createdAt;

  factory WorkOrder.fromJson(Map<String, dynamic> json) => WorkOrder(
        id: json['id'] as String,
        orderType: json['order_type'] as String,
        technicalLocation: TechnicalLocation.fromJson(json['technical_location'] as Map<String, dynamic>),
        assignedTechnicianId: json['assigned_technician_id'] as String,
        priority: json['priority'] as int,
        plannedStart: DateTime.parse(json['planned_start'] as String),
        plannedEnd: DateTime.parse(json['planned_end'] as String),
        status: WorkOrderStatus.values.byName(json['status'] as String),
        shiftNumber: json['shift_number'] as int,
        steps: (json['steps'] as List<dynamic>?)
                ?.map((e) => OTStep.fromJson(e as Map<String, dynamic>))
                .toList() ??
            [],
        createdById: json['created_by_id'] as String?,
        plannerGroup: json['planner_group'] as String?,
        activityClass: json['activity_class'] as String?,
        installationState: json['installation_state'] as String?,
        notifFinal: json['notif_final'] as bool? ?? false,
        sinTtbjoReal: json['sin_ttbjo_real'] as bool? ?? false,
        createdAt: json['created_at'] != null ? DateTime.parse(json['created_at'] as String) : null,
      );
}

class TechnicianWorkload {
  const TechnicianWorkload({
    required this.technicianId,
    required this.technicianName,
    required this.otCount,
    required this.shiftNumber,
  });

  final String technicianId;
  final String technicianName;
  final int otCount;
  final int shiftNumber;

  factory TechnicianWorkload.fromJson(Map<String, dynamic> json) =>
      TechnicianWorkload(
        technicianId: json['technician_id'] as String,
        technicianName: json['technician_name'] as String,
        otCount: json['ot_count'] as int,
        shiftNumber: json['shift_number'] as int,
      );
}
